### ISSUE TITLE
Project structure

### ISSUE THREAD

**Issue by LeMinaw** [Julien00859/pip-init#2]
I started [this wiki page](../wiki/Project-structure) for discussing project structure.

As you might see it's still incomplete and some questions have to be discussed :) feel free to give feedback.

We can add another section to discuss opt-ins tools structure if needed.

**Comment by Julien00859** [comment]
Regarding the few questions you listed in the "Open Questions" sections

> * No setup.py?

A modern pip is capable of detecting a `setup.cfg` and to install an application without running a `setup.py`. Older pip can have troubles doing it but setuptools is `setup.cfg` (always?) aware. A setup-cfg compatible setup-py file only contains two statements : `import setuptools; setuptools.setup()` and uses the setup-cfg file.

I recommend we don't include a setup.py file as modern pip are setup.cfg compatible.

> *  What should we do about setup.cfg?

My understanding on pyproject.toml vs setup.cfg is that pyproject.toml is used to declare the build tool and setup.cfg is the actual build configuration.

> * How to include license file in wheels?
> * Where do data files (eg. images) lie? How to include them?

See prior work here <https://github.com/readthedocs-fr/bin-server/commit/e40681e49f8557060c6d4c2e82a759fa7c44d503>

**Comment by Nolent** [comment]
Here the [official documentation](https://docs.python.org/3.9/distutils/setupscript.html#installing-additional-files) about data files. It's not very helping...

**Comment by LeMinaw** [comment]
> I recommend we don't include a setup.py file as modern pip are setup.cfg compatible.
Agreed, but it seems @fblanchetNaN wants to include a `setup.py` by default when using `setuptools`?

>My understanding on pyproject.toml vs setup.cfg is that pyproject.toml is used to declare the build tool and setup.cfg is the actual build configuration.
Yeah, that's basically what I understood too and what therefore what I suggested.

On the data files topic, I think it's nice to include an `assets/` folder (or whichever name we want it to be), added to the TOML so we don't have to include a MANIFEST.

**Comment by LeMinaw** [comment]
Yeah, I remember you explaining that earlier.

So, a bare `setup.py` containing
```py
import setuptools

setuptools.setup()
```
is enough to support editable installs?

If that is the case it may be worth to include it by default -maybe with a comment describing why it's there in it.

**Comment by Nolent** [comment]
Why create by default a file that will be use by a minority of person ? We should use an option for that

**Comment by fblanchetNaN** [comment]
Actually when you develop a project you should use editable install, it creates symlink in site-package to your local working directory. Even if it could work without it because lookup resolution also take into account your working directory, it seems it can produces unexpected behaviors -- I personally never faced them, but Lazor did, I think.

IMHO, as incipyt is dedicated for developers, it seems legit to install what you are supposed to use.

**Comment by Julien00859** [comment]
I 100% agree with Avanini (could you guys update your profile picture so it matches discord ?), this is the way I setup all my projects:

```sh
$ git init someproject
$ cd someproject
$ mkdir someproject
$ touch someproject/__init__.py
$ echo "print('Hello')" > someproject/__main__.py
$ vim setup.cfg  # the casual stuff
$ cat > setup.py <<EOF
import setuptools
setuptools.setup()
EOF
$ python3 -m venv venv
$ venv/bin/pip install -e .
$ venv/bin/python -m someproject
```

Using the developer mode, I can use an absolute import hook for `someproject` e.g. `import someproject` works anywhere in the source code including the tests. I never use *relative import* as I argue they are both error prone in development and do not work the way the developer expect them when you distribute the package. Basically you are forced to distribute a docker container and to write the shell line that starts your project.

Working in developer mode, I'm sure the import system works the same both in development in my machine *and* in production after a `pip3 install someproject`. In both situation you just `python3 -m someproject` and you have "someproject" available as absolute import.

**Comment by LeMinaw** [comment]
I edited the wiki page to reflect those decisions.

Does it look good to everyone?
